Node/src/node/mod.rs

```diff
@@ -98,19 +113,97 @@ impl Node {
-        if pending_tx_lists.is_empty() {
+        if submit {
+            if let Some(batch) = self.batch_to_send.take() {
+                self.ethereum_l1
+                    .execution_layer
+                    .send_batch_to_l1(
+                        batch.l2_blocks,
+                        batch.anchor_origin_height,
+                        batch.timestamp_sec,
+                    )
+                    .await?;
+                self.batch_to_send = None;
+            }
+        } else {
+            if self.batch_to_send.is_some() {
+                return Ok(()); // batch ready, skip preconfirmation until submit
+            }
+        }
+
+        if let Some(pending_tx_list) = self.taiko.get_pending_l2_tx_lists_from_taiko_geth().await? {
+            let last_anchor_origin_height = self.get_last_anchor_origin_height().await?;
+            let preconfirmation_timestamp = self.get_preconfirmation_timestamp().await?;
+            let state = self
+                .batch_builder
+                .handle_l2_block(
+                    pending_tx_list.clone(),
+                    last_anchor_origin_height,
+                    preconfirmation_timestamp,
+                )
+                .await?;
+
+            match state {
+                BuilderState::InProgress => {
+                    self.taiko
+                        .advance_head_to_new_l2_block(pending_tx_list, last_anchor_origin_height)
+                        .await?;
+                }
+                BuilderState::BatchCapacityFull(batch) => {
+                    // cannot advance the head because new batch will be created with new anchor origin height after submitting the current batch
+                    self.submit_or_keep_batch(batch, submit).await?;
+                }
+                BuilderState::MaxBlocksPerBatch(batch) => {
+                    self.taiko
+                        .advance_head_to_new_l2_block(pending_tx_list, last_anchor_origin_height)
+                        .await?;
+                    self.submit_or_keep_batch(batch, submit).await?;
+                }
+            }
+        } else {
             debug!("No pending txs, skipping preconfirmation");
-            return Ok(());
         }
 
-        self.taiko
-            .advance_head_to_new_l2_blocks(pending_tx_lists.clone())
+        Ok(())
+    }
+
+    async fn submit_or_keep_batch(&mut self, batch: Batch, submit: bool) -> Result<(), Error> {
+        if submit {
+            self.ethereum_l1
+                .execution_layer
+                .send_batch_to_l1(
+                    batch.l2_blocks,
+                    batch.anchor_origin_height,
+                    batch.timestamp_sec,
+                )
+                .await?;
+        } else {
+            self.batch_to_send = Some(batch);
+        }
+
+        Ok(())
+    }
+
+    async fn get_last_anchor_origin_height(&self) -> Result<u64, Error> {
```

Comment:
Nit: We are not getting the "last anchor origin" here, but the anchor origin to set in the newly proposing block (which references the last origin when calculating). Also, "anchor block id" is the term used in the contracts. So how about:

```suggestion
    async fn get_anchor_block_id(&self) -> Result<u64, Error> {
```